### ISSUE TITLE
Segfault at runtime

### ISSUE THREAD

**Issue by certik** [lcompilers/lpython#862]
This code:
```python
from ltypes import i32, i64
from numpy import empty, int32

def test_list(n: i32) -> i64:
    a: i32[n] = empty(n, dtype=int32)
    i: i32
    for i in range(n):
        a[i] = i
    s: i64 = 0
    for i in range(n):
        s += a[i]
    return s

print(test_list(100000001))
```
Compiles, but segfaults at runtime:
```console
$ lpython b.py
$ ./a.out 
zsh: segmentation fault  ./a.out
```

**Comment by certik** [comment]
It works for this number `1000001` but fails for a higher one. I've seen this before, this is typically caused by using `alloca` inside the loop, which runs out of stack. One must allocate any temporaries outside the loop.

**Comment by Thirumalai-Shaktivel** [comment]
Yup, I have seen this on Quick Sort benchmark.

**Comment by certik** [comment]
Here are the two loop bodies (with `--fast` to simplify, it still segfaults):
```llvm
loop.body:                                        ; preds = %.entry, %loop.body
  %4 = phi i32 [ %7, %loop.body ], [ 0, %.entry ]
  %5 = sext i32 %4 to i64
  %6 = getelementptr inbounds i32, i32* %2, i64 %5
  store i32 %4, i32* %6, align 4
  %7 = add i32 %4, 1
  %.not = icmp sgt i32 %7, %3
  br i1 %.not, label %loop.head1.preheader, label %loop.body

loop.body2:                                       ; preds = %loop.head1.preheader, %loop.body2
  %8 = phi i32 [ %14, %loop.body2 ], [ 0, %loop.head1.preheader ]
  %s.020 = phi i64 [ %13, %loop.body2 ], [ 0, %loop.head1.preheader ]
  %9 = sext i32 %8 to i64
  %10 = getelementptr inbounds i32, i32* %2, i64 %9
  %11 = load i32, i32* %10, align 4
  %12 = sext i32 %11 to i64
  %13 = add i64 %s.020, %12
  %14 = add i32 %8, 1
  %.not18 = icmp sgt i32 %14, %3
  br i1 %.not18, label %loop.end3, label %loop.body2
```
I don't actually see anything wrong with this, no alloca. But it must be something like that.

**Comment by certik** [comment]
I know what the issue is: it allocates the whole local array on stack:
```llvm
define i64 @test_list(i32* nocapture readonly %n) local_unnamed_addr #0 {
.entry:
  %0 = load i32, i32* %n, align 4
  %1 = zext i32 %0 to i64
  %2 = alloca i32, i64 %1, align 4
```
Which is fine for smaller arrays, but this one is too large so it segfaults. I can't really increase my stack on macOS. We must allocate this on a heap, possibly configure it with a compiler option.

**Comment by czgdp1807** [comment]
Yeah, I faced this issue a lot many times in the past. Basically because of using large arrays or a bunch of arrays whose total memory usage sum up to a large number. May be we can do always on heap option. NumPy allocates arrays on heap all the time I think.

**Comment by certik** [comment]
Yes. For local arrays we can have a custom heap "stack allocator" that just shifts a pointer, and shifts it back when the function returns.

**Comment by czgdp1807** [comment]
So I and @certik had a discussion on how we can fix this. I am listing those down from short term to long term approaches,

1. **Add a compiler option to choose from heap or stack memory** - This is very simple approach, just add an option whether to use stack or heap memory for a user's program. If you specify heap then all your arrays will be allocated on heap memory, similarly for stack. Cons - slow `malloc` calls, memory fragmentation.
2. **Emulate stack on heap** - Here the user will specify that heap memory should be used but now they will also specify the maximum amount of heap memory their program is supposed to be using. Say they provide 3GB. Now at the start of execution, a big 3GB will be allocated on heap memory and then it will be used as a stack. Allocation will mean shift the pointer up and freeing would mean shift the pointer down.
3. **A mix of heap and stack memory** - Here there will be two stacks, one naturally available and the other emulated on heap by us. A mix of two basically. If natural stack goes out, then heap's stack will be used.
4. **Profiler guided allocation** - Here compiler will figure out which arrays use large amounts of memory then make `malloc` calls only for those.
5. An easy option is to try extending the natural stack by OS utilities.

**Comment by certik** [comment]
Yes. I will note that the steps 1 - 5 are for local (automatic) arrays, like the one in this issue. Then there are allocatable arrays, which can be an array that is a local array in the main program or in some procedure, but then it gets allocated somewhere deep in the stack --- such arrays much be allocated on a heap, separate from the stack or "emulated stack". However, even in this case, the compiler might create some special handling, since we now which arrays and where they can be allocated and deallocated (i.e., it's not random).
